LinkControl: avoid showing "Recently updated" when there are no recently updated Pages from the API

getdave opened the pull request:
## Description
This PR fixes an error in the `<LinkControl>` component whereby the "Recently updated" label would still be displayed even if there were no recently updated Pages returned by the API request.

Now if the site in question has no published Pages, the "Recently updated" label is never shown and no results are shown. 

This bug was uncovered by changes to e2e tests in https://github.com/WordPress/gutenberg/pull/18869#issuecomment-636931681 which caused the bug to trigger an Axe warning in the tests. 

## How has this been tested?

This change manually tested (see instructions below) and is covered by new unit tests.

## Testing Instructions

#### On Master

* Delete all existing `Pages` from your website.
* Create a new Post.
* Add a Navigation block.
* Click the "Create empty" option to create a blank navigation.
* See Link UI (`LinkControl`) is displayed.
* See "Recently updated" label below search input is displayed but no results are shown.

#### On this PR's branch

* Delete all existing `Pages` from your website.
* Create a new Post.
* Add a Navigation block.
* Click the "Create empty" option to create a blank navigation.
* See Link UI (`LinkControl`) is displayed.
* See that the "Recently updated" label below search input is **_NOT_** displayed.
* Check that `aria-expanded` attribute on search input is set to `false` (also covered by tests).

## Screenshots <!-- if applicable -->

#### Broken state example 

The following shows the broken state. This is fixed by this PR.
![Screen Shot 2020-06-03 at 13 44 31](https://user-images.githubusercontent.com/444434/83638051-797ff200-a5a0-11ea-90f2-baf138119693.png)

#### After fix in this PR

![Screen Shot 2020-06-03 at 13 53 06](https://user-images.githubusercontent.com/444434/83638783-9832b880-a5a1-11ea-973d-b7007e5d6b61.png)


## Types of changes
Bug fix (non-breaking change which fixes an issue)
## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/master/docs/contributors/native-mobile.md -->

getdave committed:
Add failing unit test to cover scenario

getdave committed:
Fix showing Recently update when there are no recently update Pages in results

shaunandrews commented:
Nice. I noticed this one last night. This works as expected.

jeryj commented:
Found one issue with the HTML in the new state. The `combobox` needs to have `aria-expanded="false"` here, as there is no `role="listbox"` present. `aria-expanded="true"` should only be present when there's also a `listbox` present.

<img width="1274" alt="Screen Shot 2020-06-03 at 9 32 49 AM" src="https://user-images.githubusercontent.com/967608/83650720-43ca1180-a57e-11ea-96bb-1afcea6968b5.png">

jeryj commented:
This is a potential fix for it. I'm not sure of if it will address all instances though:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/url-input/index.js#L439

Changing that from `aria-expanded={ showSuggestions }` to `aria-expanded={ showSuggestions && !! suggestions.length }`.

getdave committed:
Update test to catch aria-expanded bug

See https://github.com/WordPress/gutenberg/pull/22856#issuecomment-638243446

getdave committed:
Fix aria-expanded a11y issue when there are no search results

Addresses https://github.com/WordPress/gutenberg/pull/22856#issuecomment-638243446

jeryj commented:
One other state I happened to notice. "Recently Update" shows up when deleting the text from a search, even if there are no pages. Watch for the end of the gif after deleting the search text to see "Recently Updated" pop up again.

![link-control-recently-updated-on-delete](https://user-images.githubusercontent.com/967608/83654519-da98cd00-a582-11ea-86d1-8e93422169f3.gif)

Sorry about the disappearing white background on `LinkControl` in the gif... No idea what happened on the quality there.

getdave committed:
Adds additional checks for aria-expanded state to help avoid future regressions

jeryj commented on packages/block-editor/src/components/url-input/index.js:
```diff
@@ -173,6 +172,7 @@ class URLInput extends Component {
 				this.setState( {
 					suggestions,
 					loading: false,
+					showSuggestions: !! suggestions.length,
```
There are a few places that check for both `showSuggestions && !! suggestions.length`. If `showSuggestions` is now tied to this, maybe we should modify those checks as well. Here are two I found:

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/url-input/index.js#L453-L454

https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/url-input/index.js#L468-L469

getdave replied:
I'd rather not touch too much about this component in this PR. I'm thinking of raising another followup Issue to track the potential need to remove `showSuggestions` entirely. It's effectively derived state and that is somewhat of an anti-pattern. We should be able to use a check of `suggestions.length` to determine whether or not to show the DOM elements. What do you think?

jeryj replied:
Sounds good!

getdave replied:
See https://github.com/WordPress/gutenberg/issues/22862

getdave commented:
Thanks for spotting these @jeryj.

> Found one issue with the HTML in the new state. The `combobox` needs to have `aria-expanded="false"`

✅ Fixed - was due to setting `showSuggestions` too early in `URLInput`.

> One other state I happened to notice. "Recently Update" shows up when deleting the text from a search, even if there are no pages. Watch for the end of the gif after deleting the search text to see "Recently Updated" pop up again.

✅ Also pushed a fix for this. Was again due to `showSuggestions` being marked as `true` too early in the API request cycle.

The above is now covered by unit tests but needs manual testing.

jeryj reviewed (approved):
Looks great! Thanks for addressing all the concerns and raising a new issue for follow-up. I manually tested with and without Pages and ran the unit tests. All states appear to be working well now! 🎉  Great work!
